Includes Memfault in the AOSP build.

diff --git a/core/tasks/vendor_module_check.mk b/core/tasks/vendor_module_check.mk
index ae967c686..15ec8d2de 100644
--- a/core/tasks/vendor_module_check.mk
+++ b/core/tasks/vendor_module_check.mk
@@ -30,6 +30,7 @@ _vendor_owner_whitelist := \
         invensense \
         intel \
         lge \
+        memfault \
         moto \
         mtk \
         nvidia \
diff --git a/target/product/base.mk b/target/product/base.mk
index 9e3862379..9a433ba9c 100644
--- a/target/product/base.mk
+++ b/target/product/base.mk
@@ -164,3 +164,7 @@ PRODUCT_COPY_FILES += $(call add-to-product-copy-files-if-exists,\
     frameworks/base/config/dirty-image-objects:system/etc/dirty-image-objects)

 $(call inherit-product, $(SRC_TARGET_DIR)/product/embedded.mk)
+
+# Memfault SDK patch version
+PRODUCT_PROPERTY_OVERRIDES += \
+    vendor.memfault.bort.version.patch=manually_patched
