diff --git a/shared/BoardConfig.mk b/shared/BoardConfig.mk
index 07f5da1..b963a8f 100644
--- a/shared/BoardConfig.mk
+++ b/shared/BoardConfig.mk
@@ -174,3 +174,6 @@ endif
 # To see full logs from init, disable ratelimiting.
 # The default is 5 messages per second amortized, with a burst of up to 10.
 BOARD_KERNEL_CMDLINE += printk.devkmsg=on
+
+include packages/apps/bort/BoardConfig.mk
+
diff --git a/shared/phone/device.mk b/shared/phone/device.mk
index 115f80f..b958713 100644
--- a/shared/phone/device.mk
+++ b/shared/phone/device.mk
@@ -51,3 +51,6 @@ ifeq ($(TARGET_USE_DYNAMIC_PARTITIONS),true)
 else
   TARGET_BUILD_SYSTEM_ROOT_IMAGE ?= true
 endif
+
+include packages/apps/bort/product.mk
+
