diff --git a/shared/BoardConfig.mk b/shared/BoardConfig.mk
index 07f5da1..b963a8f 100644
--- a/shared/BoardConfig.mk
+++ b/shared/BoardConfig.mk
@@ -174,3 +174,6 @@ endif
 # To see full logs from init, disable ratelimiting.
 # The default is 5 messages per second amortized, with a burst of up to 10.
 BOARD_KERNEL_CMDLINE += printk.devkmsg=on
+
+include vendor/memfault/bort/BoardConfig.mk
+
diff --git a/shared/device.mk b/shared/device.mk
index 3f1183e..d758657 100644
--- a/shared/device.mk
+++ b/shared/device.mk
@@ -51,6 +51,9 @@ PRODUCT_PROPERTY_OVERRIDES += \
 PRODUCT_PROPERTY_OVERRIDES += \
     wlan.driver.status=ok

+# Enforce privapp-permissions whitelist.
+PRODUCT_PROPERTY_OVERRIDES += ro.control_privapp_permissions=enforce
+
 #
 # Packages for various cuttlefish-specific tests
 #
diff --git a/shared/phone/device.mk b/shared/phone/device.mk
index 115f80f..b958713 100644
--- a/shared/phone/device.mk
+++ b/shared/phone/device.mk
@@ -51,3 +51,6 @@ ifeq ($(TARGET_USE_DYNAMIC_PARTITIONS),true)
 else
   TARGET_BUILD_SYSTEM_ROOT_IMAGE ?= true
 endif
+
+include vendor/memfault/bort/product.mk
+
