diff --git a/shared/BoardConfig.mk b/shared/BoardConfig.mk
index 7c5f6ed..9a54868 100644
--- a/shared/BoardConfig.mk
+++ b/shared/BoardConfig.mk
@@ -188,3 +188,5 @@ BOARD_USES_RECOVERY_AS_BOOT := true
 BOARD_MKBOOTIMG_ARGS += --header_version $(BOARD_BOOT_HEADER_VERSION)
 PRODUCT_COPY_FILES += device/google/cuttlefish/dtb.img:dtb.img
 BOARD_BUILD_SYSTEM_ROOT_IMAGE := false
+
+include vendor/memfault/bort/BoardConfig.mk
diff --git a/shared/device.mk b/shared/device.mk
index 72e27d7..3e1cbf8 100644
--- a/shared/device.mk
+++ b/shared/device.mk
@@ -93,6 +93,9 @@ PRODUCT_PROPERTY_OVERRIDES += drm.service.enabled=true
 PRODUCT_SOONG_NAMESPACES += hardware/google/camera
 PRODUCT_SOONG_NAMESPACES += hardware/google/camera/devices/EmulatedCamera
 
+# Enforce privapp-permissions whitelist.
+PRODUCT_PROPERTY_OVERRIDES += ro.control_privapp_permissions=enforce
+
 #
 # Packages for various GCE-specific utilities
 #
@@ -466,3 +469,5 @@ PRODUCT_SOONG_NAMESPACES += external/mesa3d
 # Need this so that the application's loop on reading input can be synchronized
 # with HW VSYNC
 PRODUCT_DEFAULT_PROPERTY_OVERRIDES += ro.surface_flinger.running_without_sync_framework=true
+
+include vendor/memfault/bort/product.mk
