diff --git a/shared/BoardConfig.mk b/shared/BoardConfig.mk
index d2907de..e04478a 100644
--- a/shared/BoardConfig.mk
+++ b/shared/BoardConfig.mk
@@ -159,3 +159,6 @@ TARGET_RECOVERY_FSTAB ?= device/google/cuttlefish/shared/config/fstab.recovery
 # To see full logs from init, disable ratelimiting.
 # The default is 5 messages per second amortized, with a burst of up to 10.
 BOARD_KERNEL_CMDLINE += printk.devkmsg=on
+
+include packages/apps/bort/BoardConfig.mk
+
diff --git a/shared/phone/device.mk b/shared/phone/device.mk
index d49d8a7..ad04b8c 100644
--- a/shared/phone/device.mk
+++ b/shared/phone/device.mk
@@ -43,3 +43,6 @@ PRODUCT_COPY_FILES += \
 # These flags are important for the GSI, but break auto
 PRODUCT_ENFORCE_RRO_TARGETS := framework-res
 PRODUCT_ENFORCE_RRO_EXCLUDED_OVERLAYS := device/google/cuttlefish/shared/overlay
+
+include packages/apps/bort/product.mk
+
