diff --git a/core/java/com/android/internal/os/BatteryStatsImpl.java b/core/java/com/android/internal/os/BatteryStatsImpl.java
index f28315bd..600ee1a4 100644
--- a/core/java/com/android/internal/os/BatteryStatsImpl.java
+++ b/core/java/com/android/internal/os/BatteryStatsImpl.java
@@ -9883,7 +9883,10 @@ public class BatteryStatsImpl extends BatteryStats {
         if (!onBattery && status == BatteryManager.BATTERY_STATUS_FULL) {
             // We don't record history while we are plugged in and fully charged.
             // The next time we are unplugged, history will be cleared.
-            mRecordingHistory = DEBUG;
+            // mRecordingHistory = DEBUG;
+
+           // For Memfault Bort's data collection, keep recording history:
+           mRecordingHistory = true;
         }
     }
